### PR TITLE
19 add code verification module

### LINKED ISSUE #19: Add Code Verification Module
V&V is composed of multiple parts, including code verification. As part of an effort to provide a comprehensive source for V&V, a code verification module should be added which allows for analysis of code verification activities

### PR DESCRIPTION
**Pull request by jrweinmeister**
This pull request adds a code verification module to CFDverify. The major changes are:

- Added `code` module
- Added unit tests for module
- Added documentation
- Added code verification tutorial for new class

The `code` module at this time only includes the `OrderOfAccuracy` class which implements methods for computing order of accuracy studies, which are the preferred code verification method.  The current class implements basic functionality plus a few utilities to export results and create plots. 

Documentation is also added, including the addition of a theory guide in the documentation. 

Changes should be consistent with the rules set out in the CFDverify development process: https://cfd-verify.readthedocs.io/en/latest/development.html

### CODE REVIEW

File: src/cfdverify/code.py [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+            fig.savefig(figure_title, bbox_inches="tight", dpi=300)
+
+
+    def plot_orders(
+            self,
+            keys : Iterable | None=None,
+            xlabel: str | None=None,
+            ylabel: str | None=None,
+            title: str | None=None,
+            save_figure: str | os.PathLike | bool | None=None,
```
Comment (wgurecky) [inline src/cfdverify/code.py:391]: I'm unsure if supporting passing None or False to these plot helper methods is ideal.  Currently, the logic would suggest no figure is saved w these inputs, so the func becomes a null op.  The only exception would be if you returned the fig object, which could be further modified by the user (like matlab's "hold").  But since you are creating a new fig and ax pair when this function is called, I don't think that use case is what you had in mind.  I suggest that you make this a required input str arg.   This would remove one path to unexpected behavior / makes the API clearer imo.

File: docs/source/theory.rst [hunk 2]
```diff
@@ -0,0 +1,42 @@
+******
+Theory
+******
+
+This section includes a minimum theory to help users perform effective verification and validation. It is not a substitute for proper training on the topic. To start, several terms can be helpful to know. Verification activities evaluate the correctness of a model. Validation activities assess a model's representation of a real system. Together, these activities are used to assess the accuracy of a model and are often talked about together as verification and validation (V&V). V&V is often part of uncertainty quantification (UQ) activities, which may include uncertainty propagation and sensitivity analysis among many other types. These activities are often described together as VVUQ.
```
Comment (dpsanjaya) [inline docs/source/theory.rst]: You can improve the writing:
This section presents a brief theory to help users perform effective verification and validation. It is not a substitute for proper training, but it is helpful for first-time users. ... Altogether, these activities are often referred to as VVUQ.

File: docs/source/theory.rst [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+.. note:: CFDverify makes a best effort attempt to follow the terminology defined in the *Verification, Validation, and Uncertainty Quantification Terminology in Computational Modeling and Simulation* standard published by the American Society of Mechanical Engineers (ASME) [ASME1]_. If you notice any discrepancies, please open an `issue <https://github.com/ORNL/cfd-verify/issues>`_.
+
+.. _code_verification_theory:
+
+Code Verification
+=================
+
+Code verification activities determine if a code is implemented correctly. There are multiple methods with varying levels of rigor that can be used to ensure a code is correct. Books which cover code verification in detail include those by Roache [Roache1998]_ or Oberkampf & Roy [Oberkampf2010]_.
```
Comment (dpsanjaya) [inline docs/source/theory.rst:14]: Books that cover code verification in detail include those written by ... and ...

File: docs/source/theory.rst [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+.. _code_verification_theory:
+
+Code Verification
+=================
+
+Code verification activities determine if a code is implemented correctly. There are multiple methods with varying levels of rigor that can be used to ensure a code is correct. Books which cover code verification in detail include those by Roache [Roache1998]_ or Oberkampf & Roy [Oberkampf2010]_. 
+
+While code developers often conduct code verification to ensure their solver is bug-free, it is ultimately the responsibility of an analyst to verify their results. When available, analysts are encouraged to run verification suites on the same executable as their analysis. The test problems should also be reviewed to ensure they contain the same equations as the analysis problem, including boundary conditions, source terms, etc. when appropriate.
```
Comment (dpsanjaya) [inline docs/source/theory.rst]: Perhaps "When possible" instead of "When available"?

As standard practices, the test problems should also contain the same equations as the analysis problem, including boundary conditions, source terms, etc. Hence, please review the test problems and make sure that they reflect your needs.

File: docs/source/theory.rst [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+Order of Accuracy
+-----------------
+
+Order of accuracy tests are the most rigorous code verification activity because they assess that the solver not only converges to the right value, but that it also converges at the formal order of accuracy. These tests often require solving the test problem on multiple meshes to ensure that the formal order is approached. It is common to not recover the *precise* formal order, but it is often sufficient to show that it is being asymptotically approached with refinement. This can require solving more than the minimum two discretization levels to conduct this test. Order of accuracy tests should be conducted with exact solutions for them best assurance; see the following two subsections for more details on the differences between exact and approximate solutions.
```
Comment (dpsanjaya) [inline docs/source/theory.rst]: Instead of "the formal order is approached", maybe "... is realized" or "... is observed"?

In general, three or more discretization level are necessary to conduct this test. 

"Their" best assurance?

**Review by dpsanjaya — COMMENTED** [review]
The code works well. I did notice that "Extrapolated Value" should be "Extrapolated value" to be consistent with "Fine mesh error" and "Fine mesh uncertainty".

File: docs/source/theory.rst (hunk 3 above)
Comment (jrweinmeister) [reply docs/source/theory.rst:14]: Addressed

File: docs/source/theory.rst (hunk 5 above)
Comment (jrweinmeister) [reply docs/source/theory.rst]: 1. I took out approach as I think it repeats information from the next sentence. The new sentence is better in that it simply states what is needed.
2. For cases with exact answers, only two discretization levels are necessary. It seems standard practice to use more, but I think it would be unwise to over-comment on standard practice without reference to a good reason.
3. Fixed

File: docs/source/theory.rst (hunk 4 above)
Comment (jrweinmeister) [reply docs/source/theory.rst]: Addressed

File: docs/source/theory.rst (hunk 2 above)
Comment (jrweinmeister) [reply docs/source/theory.rst]: Addressed. I expect this section to be expanded later as I add details on solution verification and validation, so I don't think it is effective to spend too much time improving this paragraph at the moment.

For the bigger picture, I think it will be important to review what the right level of depth to include in the code's theory section is.

File: src/cfdverify/code.py (hunk 1 above)
Comment (jrweinmeister) [reply src/cfdverify/code.py:391]: I fixed this, thanks for catching me making it complicated. Additionally, I am opening an issue (#26) to make my helper functions more consistent with other python libraries so that the interface is as expected.
